### PR TITLE
Add kyber staking and modify voting power strategy

### PR DESCRIPTION
**Pull request by ducquangkstn**

### CODE REVIEW

File: contracts/governance/KyberStaking.sol [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+  ) PermissionAdmin(admin) {
+    require(_epochPeriod > 0, 'ctor: epoch period is 0');
+    require(_kncToken != IERC20(0), 'ctor: kncToken 0');
+
+    epochPeriodInSeconds = _epochPeriod;
+    firstEpochStartTimestamp = _startTimestamp;
+    kncToken = _kncToken;
+  }
+
+  function updateWithdrawHandlers(IWithdrawHandler _withdrawHandler) external onlyAdmin {
```
Comment (manhlx3006) [inline contracts/governance/KyberStaking.sol:64]: `updateWithdrawHandlers` -> `updateWithdrawHandler`

File: contracts/governance/KyberStaking.sol [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+    delegatedStake = stakerData.delegatedStake;
+    representative = stakerData.representative;
+  }
+
+  /**
+   * @notice return raw data of a staker for an epoch
+   *         WARN: should be used only for initialized data
+   *          if data has not been initialized, it will return all 0
+   *          pool master shouldn't use this function to compute/distribute rewards of pool members
+   * @dev  in KyberDao contract, if staker wants to claim reward for past epoch,
```
Comment (manhlx3006) [inline contracts/governance/KyberStaking.sol:224]: Please change the comments, some of them are related to the old one

File: contracts/governance/KyberVotingPowerStrategy.sol [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    /// init epochutils
+    epochPeriodInSeconds = _staking.epochPeriodInSeconds();
+    firstEpochStartTimestamp = _staking.firstEpochStartTimestamp();
+  }
+
+  /// @dev endTimestamp: furture usage
+  function handleProposalCreation(
+    uint256 proposalID,
+    uint256 startTimestamp,
```
Comment (manhlx3006) [inline contracts/governance/KyberVotingPowerStrategy.sol:42]: should change `proposalID` -> `proposalId` and `startTimestamp` -> `startTime` for consistency

File: contracts/governance/KyberVotingPowerStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    uint256 /*choice*/
+  ) external override returns (uint256 votingPower) {
+    require(msg.sender == address(governance), 'not governance');
+
+    (uint256 stake, uint256 dStake, address representative) =
+      staking.initAndReturnStakerDataForCurrentEpoch(voter);
+    return representative == voter ? stake.add(dStake) : dStake;
+  }
+
+  function handleWithdraw(address user, uint256 /*reduceAmount*/) external override {
```
Comment (manhlx3006) [inline contracts/governance/KyberVotingPowerStrategy.sol:85]: should be called only by KyberStaking contract

File: contracts/governance/KyberVotingPowerStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    firstEpochStartTimestamp = _staking.firstEpochStartTimestamp();
+  }
+
+  /// @dev endTimestamp: furture usage
+  function handleProposalCreation(
+    uint256 proposalID,
+    uint256 startTimestamp,
+    uint256 /*endTimestamp*/
+  ) external override {
+    require(msg.sender == address(governance), 'not governance');
```
Comment (manhlx3006) [inline contracts/governance/KyberVotingPowerStrategy.sol:45]: since it is used in many func, consider to use modifer

**Review by manhlx3006 — COMMENTED** [review]
Only review 3 files KyberStaking, IVotingPowerStrategy and KyberVotingPowerStrategy,
please mark the PR as ready when it is ok to review and merge

File: contracts/governance/KyberStaking.sol (hunk 1 above)
Comment (ducquangkstn) [reply contracts/governance/KyberStaking.sol]: change as request

File: contracts/governance/KyberStaking.sol (hunk 2 above)
Comment (ducquangkstn) [reply contracts/governance/KyberStaking.sol]: change as request

File: contracts/governance/KyberVotingPowerStrategy.sol (hunk 3 above)
Comment (ducquangkstn) [reply contracts/governance/KyberVotingPowerStrategy.sol]: change as request

File: contracts/governance/KyberVotingPowerStrategy.sol (hunk 4 above)
Comment (ducquangkstn) [reply contracts/governance/KyberVotingPowerStrategy.sol]: change as request

File: contracts/governance/KyberVotingPowerStrategy.sol (hunk 5 above)
Comment (ducquangkstn) [reply contracts/governance/KyberVotingPowerStrategy.sol]: change as request

File: contracts/governance/KyberStaking.sol [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+interface IWithdrawHandler {
+  function handleWithdrawal(address staker, uint256 reduceAmount) external;
+}
+
+/**
+ * @notice   This contract is using SafeMath for uint, which is inherited from EpochUtils
+ *           Some events are moved to interface, easier for public uses
+ */
+contract KyberStaking is IKyberStaking, EpochUtils, ReentrancyGuard, PermissionAdmin {
+  using Math for uint256;
```
Comment (manhlx3006) [inline contracts/governance/KyberStaking.sol:22]: why do we need both Math and SafeMath here?

File: contracts/governance/KyberStaking.sol [hunk 7]
```diff
@@ -0,0 +1,466 @@
...
+  using SafeMath for uint256;
+  /// TODO: should we compress uint128-uint128-address-bool
```
Comment (manhlx3006) [inline contracts/governance/KyberStaking.sol:24]: We should be able to use uint128-uin128 for storing stake & delegated stake

**Review by manhlx3006 — COMMENTED** [review]
Overall looks good to me

File: contracts/governance/KyberStaking.sol (hunk 7 above)
Comment (ducquangkstn) [reply contracts/governance/KyberStaking.sol:24]: Yeah, I mean we should do it later. For now, just copy code from old repo for drafting

File: contracts/governance/KyberStaking.sol (hunk 6 above)
Comment (ducquangkstn) [reply contracts/governance/KyberStaking.sol:22]: Math for `min` function
SafeMath for `add` `sub` `mul` `divide`

File: contracts/governance/KyberVotingPowerStrategy.sol [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    address user,
+    uint256 /*reduceAmount*/
+  ) external override onlyStaking {
+    uint256 currentEpoch = getCurrentEpochNumber();
+    (uint256 stake, uint256 dStake, address representative) = staking.getStakerData(
+      user,
+      currentEpoch
+    );
+    uint256 votingPower = representative == user ? stake.add(dStake) : dStake;
+    governance.handleVotingPowerChanged(user, votingPower, epochProposals[currentEpoch]);
```
Comment (manhlx3006) [inline contracts/governance/KyberVotingPowerStrategy.sol:98]: should we do callback only if `epochProposals[currentEpoch]` is not empty?

**Review by manhlx3006 — COMMENTED** [review]
Please add more comments for each functions, can do it in another PR

**Review by manhlx3006 — APPROVED** [review]

**Review by manhlx3006 — APPROVED** [review]

**Pull request merged**